### PR TITLE
Timetables: Store timetables 

### PR DESCRIPTION
**Pull request by zacharytang**
#40 In progress

**Completed**:
- Expanding of shortened URL
- Parsing of URL to retrieve information
- Usage of NUSMods API to retrieve module class information
- Parsing of class info to retrieve timings
- Storage of timings into a  #timetable
- Storage of timetables by person
- Modifying add command to take in url
- Added tests and updated current tests

**TODO**:
- Make timetable attribute optional

### CODE REVIEW

**Review by April0616 — COMMENTED** [review]
Is it a complete one?
Would you like to merge it??

File: docs/UserGuide.adoc [hunk 1]
```diff
@@ -96,11 +96,11 @@ Examples:
...
 ==== Adding timetables
 
-If you want to save a person's timetable, add their NUSMods link when adding them to the address book.
+If you want to save a person's timetable, add their shortened NUSMods link when adding them to the address book.
 Their timetable will be saved and will be available to be viewed at any time
 
-[TIP]
-Address Book accepts both full and shortened NUSMods links!
+[IMPORTANT]
+Address Book only accepts shortened NUSMods links
```
Comment (nbriannl) [inline docs/UserGuide.adoc:103]: Presume full URLs will come at a later enhancement.

File: src/main/java/seedu/address/commons/util/timetable/Lesson.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    }
+
+    public String getStartTime() {
+        return startTime;
+    }
+
+    public String getEndTime() {
+        return endTime;
+    }
+}
```
Comment (nbriannl) [inline src/main/java/seedu/address/commons/util/timetable/Lesson.java]: Standard getter methods. 👍

File: src/main/java/seedu/address/commons/util/timetable/TimetableInfoFromUrl.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+     * Adds lesson information for a specific module to the timetable information.
+     */
+    public void addModuleInfo(ModuleInfoFromUrl moduleInfoToAdd) {
+        if (moduleInfo.contains(moduleInfoToAdd)) {
+            moduleInfo.set(moduleInfo.indexOf(moduleInfoToAdd), moduleInfoToAdd);
+        } else {
+            moduleInfo.add(moduleInfoToAdd);
+        }
+    }
+}
```
Comment (nbriannl) [inline src/main/java/seedu/address/commons/util/timetable/TimetableInfoFromUrl.java]: 🔢

File: src/main/java/seedu/address/commons/util/timetable/TimetableParserUtil.java [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    private static void constructTimetableForModule(String acadYear, String semester, TimetableInfo timetable,
+                                                    ModuleInfoFromUrl moduleInfo) throws IllegalValueException {
+        ArrayList<Lesson> lessons = getLessonInfoFromApi(acadYear, semester, moduleInfo.getModCode());
+        HashMap<String, String> lessonsForModule = moduleInfo.getLessonInfo();
+
+        for (String classType : lessonsForModule.keySet()) {
+            String classNo = lessonsForModule.get(classType);
+
+            addLessonToTimetable(timetable, lessons, classType, classNo);
+        }
```
Comment (nbriannl) [inline src/main/java/seedu/address/commons/util/timetable/TimetableParserUtil.java]: 💯

File: src/main/java/seedu/address/commons/util/timetable/TimetableParserUtil.java [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    private static void addLessonToTimetable(TimetableInfo timetable, ArrayList<Lesson> lessons, String classType,
+                                             String classNo) throws IllegalValueException {
+        for (Lesson lesson : lessons) {
+            if (lessonExistsInParsedInfo(classType, classNo, lesson)) {
+                timetable.updateSlotsWithLesson(lesson.getWeekType(), lesson.getDay(), lesson.getStartTime(),
+                        lesson.getEndTime());
+            }
+        }
+    }
+
```
Comment (nbriannl) [inline src/main/java/seedu/address/commons/util/timetable/TimetableParserUtil.java]: 😨

File: src/main/java/seedu/address/commons/util/timetable/TimetableParserUtil.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+            return WEEK_ODD;
+        case "Even Week":
+            return WEEK_EVEN;
+        case "Every Week":
+            return WEEK_BOTH;
+        default:
+            throw new IllegalValueException(MESSAGE_INVALID_WEEK_TYPE);
+        }
+    }
+}
```
Comment (nbriannl) [inline src/main/java/seedu/address/commons/util/timetable/TimetableParserUtil.java]: 💀

**Review by nbriannl — APPROVED** [review]
Just write a detailed documentation of this feature in the Dev Guide will do. Thanks for the hard work This enhancement is greatly appreciated.

**Pull request merged**